sympy/solvers/tests/test_solveset.py

```diff
@@ -1724,6 +1731,150 @@ def test_solve_nonlinear_trans():
+            S(43)/2))/3)/6 + ((3*sqrt(3)*31985**(S(1)/4)*sin(atan(S(172)/49)/2)/2 +
+            S(43)/2)**2 + (27 + 3*sqrt(3)*31985**(S(1)/4)*cos(atan(S(172)/49)/2)
+            /2)**2)**(S(1)/6)*cos(atan((27 + 3*sqrt(3)*31985**(S(1)/4)*cos(atan(
+            S(172)/49)/2)/2)/(3*sqrt(3)*31985**(S(1)/4)*sin(atan(S(172)/49)/2)/2 +
+            S(43)/2))/3)/6 + I*(-sqrt(3)*((3*sqrt(3)*31985**(S(1)/4)*sin(atan(
+            S(172)/49)/2)/2 + S(43)/2)**2 + (27 + 3*sqrt(3)*31985**(S(1)/4)*cos(
+            atan(S(172)/49)/2)/2)**2)**(S(1)/6)*cos(atan((27 + 3*sqrt(3)*31985**(
+            S(1)/4)*cos(atan(S(172)/49)/2)/2)/(3*sqrt(3)*31985**(S(1)/4)*sin(
+            atan(S(172)/49)/2)/2 + S(43)/2))/3)/6 + ((3*sqrt(3)*31985**(S(1)/4)*
+            sin(atan(S(172)/49)/2)/2 + S(43)/2)**2 + (27 + 3*sqrt(3)*31985**(S(1)/4)*
+            cos(atan(S(172)/49)/2)/2)**2)**(S(1)/6)*sin(atan((27 + 3*sqrt(3)*31985**(
+            S(1)/4)*cos(atan(S(172)/49)/2)/2)/(3*sqrt(3)*31985**(S(1)/4)*sin(
+            atan(S(172)/49)/2)/2 + S(43)/2))/3)/6 - 4*im(1/((-S(1)/2 + sqrt(3)*I/2)*
+            (S(43)/2 + 27*I + sqrt(-256 + (43 + 54*I)**2)/2)**(S(1)/3)))/3))
+
+    assert solveset(n1, x) == res
+
+
+def test_issue_13961():
+    V = (ax, bx, cx, gx, jx, lx, mx, nx, q) = symbols('ax bx cx gx jx lx mx nx q')
+    S = (ax*q - lx*q - mx, ax - gx*q - lx, bx*q**2 + cx*q - jx*q - nx, q*(-ax*q + lx*q + mx), q*(-ax + gx*q + lx))
+
+    sol = FiniteSet((lx + mx/q, (-cx*q + jx*q + nx)/q**2, cx, mx/q**2, jx, lx, mx, nx, q),
+                    (lx + mx/q, (cx*q - jx*q - nx)/q**2*-1, cx, mx/q**2, jx, lx, mx, nx, q))
+    assert nonlinsolve(S, *V) == sol
+    # The two solutions are in fact identical, so even better if only one is returned
+
+
+def test_issue_14541():
+    solutions = solveset(sqrt(-x**2 - 2.0), x)
+    assert abs(solutions.args[0]+1.4142135623731*I) <= 1e-9
+    assert abs(solutions.args[1]-1.4142135623731*I) <= 1e-9
+
+
+def test_issue_13396():
+    expr = -2*y*exp(-x**2 - y**2)*Abs(x)
+    # The second part may be empty
+    sol = Union(FiniteSet(Integer(0)), Intersection(S.Reals, FiniteSet(Pow(Add(Mul(Integer(-1),
+                Pow(x, Integer(2))), zoo), Rational(1, 2)), Mul(Integer(-1),
+                 Pow(Add(Mul(Integer(-1), Pow(x, Integer(2))), zoo), Rational(1, 2))))))
+
+    # Direct comparison does not work
+    assert srepr(solveset(expr, y, domain=S.Reals)) == srepr(sol)
+    # Original issue was TypeError
+
+
+@XFAIL
+def test_issue_13396_failing():
+    expr = -2*y*exp(-x**2 - y**2)*Abs(x)
+    # The second part may be empty
+    sol = Union(FiniteSet(Integer(0)), Intersection(S.Reals, FiniteSet(Pow(Add(Mul(Integer(-1),
+                Pow(x, Integer(2))), zoo), Rational(1, 2)), Mul(Integer(-1),
+                 Pow(Add(Mul(Integer(-1), Pow(x, Integer(2))), zoo), Rational(1, 2))))))
+
+     # Should work, but will not. See #21943
+    assert solveset(expr, y, domain=S.Reals) == sol
```

Comment:
Problem is that the args are flipped in `zoo - x**2`.

Reply:
I guess that the *cause* of the problem is evaluate=False being used somewhere...

Reply:
I tried to rewrite the expressions some time ago, but couldn't figure out where to put the `evaluate=False`. May have another go, but indeed it would be better to only get one solution...

Reply:
To recreate the output of `solveset` we need `evaluate=False` for the `Add` in `zoo - x**2`:
```python
In [15]: sol = Union(FiniteSet(Integer(0)), Intersection(S.Reals, FiniteSet(Pow(Add(Mul(Integer(-1),
    ...:             Pow(x, Integer(2))), zoo, evaluate=False), Rational(1, 2)), Mul(Integer(-1),
    ...:             Pow(Add(Mul(Integer(-1), Pow(x, Integer(2))), zoo, evaluate=False), Rational(1, 2))))))
    ...: 
    ...: expr = -2*y*exp(-x**2 - y**2)*Abs(x)
    ...: 
    ...: print(solveset(expr, y, Reals) == sol)
True
```

Reply:
It's the 2-arg Mul:
```python
In [6]: e = zoo - x**2

In [7]: e.args
Out[7]: 
⎛       2⎞
⎝zoo, -x ⎠

In [8]: e2 = -1 * e

In [9]: e2
Out[9]: 
 2      
x  + zoo

In [10]: e2.args
Out[10]: 
⎛ 2     ⎞
⎝x , zoo⎠

In [11]: e2 == e2.func(*e2.args)
Out[11]: False
```
It can be fixed with
```diff
diff --git a/sympy/core/mul.py b/sympy/core/mul.py
index 9933eb601a..960b5baf25 100644
--- a/sympy/core/mul.py
+++ b/sympy/core/mul.py
@@ -281,7 +281,7 @@ def flatten(cls, seq):
                         ar = a*r
                         if ar:
                             bargs.insert(0, ar)
-                        bargs = [Add._from_args(bargs)]
+                        bargs = [Add(*bargs)]
                         rv = bargs, [], None
             if rv:
                 return rv
```

Reply:
or maybe `_unevaluated_Add(*args)` -- I read the issue but am not seeing enough context to understand the problem. `Add._from_args` is much faster than `Add` and since you are making this change in core it makes me nervous.

Reply:
> It can be fixed with

I think this is best. Otherwise it's dicey with special handling of objects in Add which get promoted to the coefficient position (like TensExpr, AccumBounds, ComplexInfinity).

And there is a difference between the handling of `-x` vs `-1*x`. The following diff will make the behavior more uniform:
```diff
diff --git a/sympy/core/mul.py b/sympy/core/mul.py
index 9933eb601a..1de76341f1 100644
--- a/sympy/core/mul.py
+++ b/sympy/core/mul.py
@@ -267,7 +267,7 @@ def flatten(cls, seq):
                 r, b = b.as_coeff_Mul()
                 if b.is_Add:
                     if r is not S.One:  # 2-arg hack
-                        # leave the Mul as a Mul?
+                        # 2*Mul(3, x + 1, evaluate=False) -> 6*(x + 1)
                         ar = a*r
                         if ar is S.One:
                             arb = b
@@ -275,14 +275,12 @@ def flatten(cls, seq):
                             arb = cls(a*r, b, evaluate=False)
                         rv = [arb], [], None
                     elif global_parameters.distribute and b.is_commutative:
-                        r, b = b.as_coeff_Add()
-                        bargs = [_keep_coeff(a, bi) for bi in Add.make_args(b)]
-                        _addsort(bargs)
+                        r, nonr = b.as_coeff_Add()
+                        bargs = [_keep_coeff(a, bi) for bi in Add.make_args(nonr)]
                         ar = a*r
                         if ar:
                             bargs.insert(0, ar)
-                        bargs = [Add._from_args(bargs)]
-                        rv = bargs, [], None
+                        rv = [Add(*bargs)], [], None
             if rv:
                 return rv
 
@@ -2144,4 +2142,4 @@ def do(e):
 
 from .numbers import Rational
 from .power import Pow
-from .add import Add, _addsort, _unevaluated_Add
+from .add import Add, _unevaluated_Add
diff --git a/sympy/core/tests/test_arit.py b/sympy/core/tests/test_arit.py
index 16580c14fa..8a35d99a49 100644
--- a/sympy/core/tests/test_arit.py
+++ b/sympy/core/tests/test_arit.py
@@ -2354,3 +2354,24 @@ def test_issue_17130():
 def test_issue_21034():
     e = -I*log((re(asin(5)) + I*im(asin(5)))/sqrt(re(asin(5))**2 + im(asin(5))**2))/pi
     assert e.round(2)
+
+
+def test_issue_22021():
+    from sympy.calculus.util import AccumBounds
+    from sympy.tensor.tensor import TensExpr
+    from sympy.utilities.iterables import permutations
+    # these objects are special cases in Mul
+    from sympy.tensor.tensor import TensorIndexType, tensor_indices, TensorSymmetry,    get_symmetric_group_sgs, TensorIndex, tensor_mul, TensAdd,tensor_heads
+    L = TensorIndexType("L")
+    i = tensor_indices("i", L)
+    A, B = tensor_heads("A B", [L])
+    e = A(i) + B(i)
+    assert -e == -1*e
+    e = zoo + x
+    assert -e == -1*e
+    a = AccumBounds(1, 2)
+    e = a + x
+    assert -e == -1*e
+    for args in permutations((zoo, a, x)):
+        e = Add(*args, evaluate=False)
+        assert -e == -1*e
```